### PR TITLE
#1757 - Create common container for content in tab

### PR DESCRIPTION
**Pull request by sh16011993**
Created a common container for contents in the tab. This ensures that there is no jumping effect when the ministry user clicks from one tab to the other ensuring a smooth UI transition. Below are the screenshots:

Profile Tab:

image

Applications tab:

image

Restrictions tab:

image

File Uploads tab:

image

SIN tab:

image

Overawards tab:

image

Notes tab: 

image

Also, checked to make sure that effects of the changes in the shared components are propagated correctly.

Student Application Summary:

image

Here, the spacing between the Applications title and the datatable is reduced (as a side effect of the above change)

### CODE REVIEW

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 1]
```diff
@@ -1,112 +1,120 @@
 <template>
   <v-card class="mb-5">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:2]: The idea was to no longer use `v-card` or `v-container`. Is it not taken care of by a common component?

File: sources/packages/web/src/components/common/StudentOverawardDetails.vue [hunk 2]
```diff
@@ -1,54 +1,58 @@
 <template>
   <v-card class="mb-5">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/StudentOverawardDetails.vue:2]: Same from https://github.com/bcgov/SIMS/pull/1816/files#r1139577805

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+export default defineComponent({
+  props: {
+    card: {
+      type: Boolean,
+      required: false,
+      default: false,
+    },
+  },
+  setup() {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:24]: The `setup` method can be removed.

File: sources/packages/web/src/components/layouts/TabContainer.vue [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      required: false,
+      default: true,
+    },
+    fullWidth: {
+      type: Boolean,
+      required: false,
+      default: false,
+    },
+  },
+  setup() {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/layouts/TabContainer.vue:25]: The `setup` method can be removed.

File: sources/packages/web/src/views/aest/student/StudentDetails.vue [hunk 5]
```diff
@@ -23,7 +23,7 @@
       </header-navigator>
     </template>
     <template #tab-header>
-      <v-tabs :model="tab" stacked color="primary"
+      <v-tabs :model="tab" stacked color="primary" class="mt-n14"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/StudentDetails.vue:26]: Can you please elaborate on why this class was needed? `class="mt-n14"`

File: sources/packages/web/src/components/generic/BodyHeader.vue [hunk 6]
```diff
@@ -1,5 +1,5 @@
 <template>
-  <v-row no-gutters>
+  <v-row no-gutters class="mb-n6">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/generic/BodyHeader.vue:2]: Can you please elaborate on why `class="mb-n6"` was needed?

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<template>
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:1]: May the template not be simplified as below or something similar?
```vue
<template>
  <v-card v-if="card">
    <slot name="header"></slot>
    <slot></slot>
  </v-card>
  <div v-else>
    <slot name="header"></slot>
    <slot></slot>
  </div>
</template>
```

File: sources/packages/web/src/components/common/notes/Notes.vue [hunk 8]
```diff
@@ -1,29 +1,28 @@
...
-            </template>
-          </check-permission-role>
-        </v-col>
-      </v-row>
-      <v-row class="m-2" v-if="!notes || notes.length === 0"
-        ><v-col
-          >No notes found. Please click on create new note to add one.</v-col
-        ></v-row
-      >
+  <v-container :fluid="true">
```
Comment (ann-aot) [inline sources/packages/web/src/components/common/notes/Notes.vue:2]: our plan was to add the new template to all tab content right?

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    <div v-else class="mt-n1"><slot></slot></div>
+  </div>
+</template>
+
+<script lang="ts">
+import { defineComponent } from "vue";
+
+export default defineComponent({
+  props: {
+    card: {
```
Comment (ann-aot) [inline sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:18]: can we give a proper name something like `isCard` or `enableCardView` or maybe something appropriate

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue [hunk 10]
```diff
@@ -0,0 +1,28 @@
+<template>
+  <div>
+    <!-- Header slot (with and without v-card) -->
+    <v-card v-if="card"><slot name="header"></slot></v-card>
+    <div v-else><slot name="header"></slot></div>
+
+    <!-- Default slot (with and without v-card) -->
+    <v-card v-if="card" class="mt-n1"><slot></slot></v-card>
```
Comment (ann-aot) [inline sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:8]: if `card==true` then will there be one `v-card` for the header and another `v-card` for the body?

File: sources/packages/web/src/components/layouts/TabContainer.vue [hunk 11]
```diff
@@ -0,0 +1,29 @@
+<template>
+  <div class="mt-8">
```
Comment (ann-aot) [inline sources/packages/web/src/components/layouts/TabContainer.vue:2]: just wanted to know, can we use `v-container` with `fluid`

File: sources/packages/web/src/components/layouts/TabContainer.vue [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+import { defineComponent } from "vue";
+
+export default defineComponent({
+  props: {
+    card: {
+      type: Boolean,
+      required: false,
+      default: true,
+    },
+    fullWidth: {
```
Comment (ann-aot) [inline sources/packages/web/src/components/layouts/TabContainer.vue:19]: looks like you are not using the prop `fullWidth`

File: sources/packages/web/src/components/layouts/TabContainer.vue [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    <slot v-else></slot>
+  </div>
+</template>
+
+<script lang="ts">
+import { defineComponent } from "vue";
+
+export default defineComponent({
+  props: {
+    card: {
```
Comment (ann-aot) [inline sources/packages/web/src/components/layouts/TabContainer.vue:14]: can we give a proper name

File: sources/packages/web/src/views/aest/student/Applications.vue [hunk 14]
```diff
@@ -1,12 +1,10 @@
...
-  <v-card class="mt-4">
-    <div class="mx-5 py-4">
-      <StudentApplications
-        :studentId="studentId"
-        @goToApplication="goToApplication"
-      />
-    </div>
-  </v-card>
+  <tab-container>
+    <StudentApplications
```
Comment (ann-aot) [inline sources/packages/web/src/views/aest/student/Applications.vue:3]: we can rename it to `student-applications`

File: sources/packages/web/src/components/common/OverawardDetails.vue (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/common/OverawardDetails.vue:2]: @sh16011993 for your ref. it must be taken care by body-header-container in this case.

File: sources/packages/web/src/components/common/notes/Notes.vue [hunk 15]
```diff
@@ -1,29 +1,28 @@
...
+    <v-row class="m-2">
+      <v-col class="category-header-large color-blue">{{ title }}</v-col>
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/notes/Notes.vue:4]: This file has formatting warnings. 

image
image

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue [hunk 16]
```diff
@@ -0,0 +1,28 @@
+<template>
+  <div>
+    <!-- Header slot (with and without v-card) -->
+    <v-card v-if="card"><slot name="header"></slot></v-card>
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:4]: I didn't get this part. The v-card or div must be at the parent/root level for both header and body right?

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue (hunk 10 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:8]: Yes, when card (enableCardView) is set to true, there is a v-card for the header and another one for the body.

**Comment by dheepak-aot** [comment]
Nice work @sh16011993 . Please take a look at the comments.

File: sources/packages/web/src/components/aest/StudentApplications.vue [hunk 17]
```diff
@@ -1,127 +1,130 @@
...
-                :disabled="sinValidStatus !== SINStatusEnum.VALID"
-                variant="plain"
-                color="primary"
-                class="label-bold"
-                @click="
-                  $emit(
-                    'editApplicationAction',
-                    slotProps.data.status,
-                    slotProps.data.id,
+        <Column
```
Comment (andrepestana-aot) [inline sources/packages/web/src/components/aest/StudentApplications.vue:25]: Not part of this PR but are we going to replace this `Column` tag as it is Primevue? @andrewsignori-aot @dheepak-aot @ann-aot @guru-aot

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue (hunk 10 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:8]: Fixed it.

File: sources/packages/web/src/components/common/OverawardDetails.vue (hunk 1 above)
Comment (sh16011993) [reply sources/packages/web/src/components/common/OverawardDetails.vue:2]: Fixed it.

File: sources/packages/web/src/components/common/StudentOverawardDetails.vue (hunk 2 above)
Comment (sh16011993) [reply sources/packages/web/src/components/common/StudentOverawardDetails.vue:2]: Fixed it.

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue (hunk 3 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:24]: Removed.

File: sources/packages/web/src/components/layouts/TabContainer.vue (hunk 4 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/TabContainer.vue:25]: Removed.

File: sources/packages/web/src/views/aest/student/StudentDetails.vue (hunk 5 above)
Comment (sh16011993) [reply sources/packages/web/src/views/aest/student/StudentDetails.vue:26]: I used this class to move the tabs up, thus reducing the space between the `full-page-container` subtitle and the `tabs-header`

File: sources/packages/web/src/components/generic/BodyHeader.vue (hunk 6 above)
Comment (sh16011993) [reply sources/packages/web/src/components/generic/BodyHeader.vue:2]: I did this change to reduce the spacing  between the title and the subtitle to match it as per the figma.

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue (hunk 7 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:1]: I have made the suggested change. There are two slight modifications. 
1) I have wrapped the card / no-card inside a div to give it a `class="mb-5"` to add `bottom-margin` after every card / no-card element. 
2) I have added a `v-container` inside the v-card

File: sources/packages/web/src/components/common/notes/Notes.vue (hunk 8 above)
Comment (sh16011993) [reply sources/packages/web/src/components/common/notes/Notes.vue:2]: Yes, fixed it.

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue (hunk 16 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:4]: Yes, I have corrected it now.

File: sources/packages/web/src/components/aest/StudentApplications.vue (hunk 17 above)
Comment (ann-aot) [reply sources/packages/web/src/components/aest/StudentApplications.vue:25]: it will be removed, when we replace primevue datable with vuetify data table

File: sources/packages/web/src/views/aest/student/StudentDetails.vue (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/aest/student/StudentDetails.vue:26]: I got the point about why it was done, I was wondering if it needed to be done with a negative margin. IMO, applying a negative can be used when there is no alternative way. For instance, maybe there is something else to be adjusted that is actually causing the additional spacing.

File: sources/packages/web/src/components/generic/BodyHeader.vue (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/generic/BodyHeader.vue:2]: Same from https://github.com/bcgov/SIMS/pull/1816#discussion_r1142728975

Comment (sh16011993) [reply sources/packages/web/src/components/generic/BodyHeader.vue:2]: Fixed this.

File: sources/packages/web/src/views/aest/student/StudentDetails.vue (hunk 5 above)
Comment (sh16011993) [reply sources/packages/web/src/views/aest/student/StudentDetails.vue:26]: For this one, `full-page-container` has `layoutTemplate` with a `div` that adds a margin `mt-4`.

Comment (andrewsignori-aot) [reply sources/packages/web/src/views/aest/student/StudentDetails.vue:26]: As we talked over the call, lets check if it is possible to have the `full-page-container` updated to the `centered-tab` template.

Comment (sh16011993) [reply sources/packages/web/src/views/aest/student/StudentDetails.vue]: Updated `full-page-container` to remove the extra padding and the margin from the CenteredTab layout and verified that it impacts the components - StudentDetails, InstitutionDetails, LocationActiveApplicationSummary and LocationCEOSummary and improves them by removing the extra white space.

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/layouts/BodyHeaderContainer.vue:1]: In terms of code syntax, I would still prefer as below for better readability.
```vue
<template>
  <div class="mb-5">
    <v-card v-if="enableCardView">
      <v-container>
        <slot name="header"></slot>
        <slot></slot>
      </v-container>
    </v-card>
    <div v-else>
      <slot name="header"></slot>
      <slot></slot>
    </div>
  </div>
</template>
```

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 18]
```diff
@@ -1,50 +1,50 @@
 <template>
-  <v-card class="mb-5">
-    <v-container fluid>
+  <body-header-container :enableCardView="true">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:2]: A general point for the devs, I have a feeling that most of the time we use this with a card. Would be the case to set the property as `true` by default? Thoughts?
@ann-aot @dheepak-aot @andrepestana-aot @guru-aot

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes.
The ticket is not precise about the ticket expectations but I was under the impression that we would be changing Student and Institution tabs as part of this effort, but I may be wrong.
@dheepak-aot @guru-aot @andrepestana-aot @ann-aot does someone recall the final goal of the ticket?

File: sources/packages/web/src/components/layouts/TabContainer.vue (hunk 13 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/TabContainer.vue]: done

File: sources/packages/web/src/components/layouts/TabContainer.vue (hunk 12 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/TabContainer.vue]: removed

File: sources/packages/web/src/components/layouts/BodyHeaderContainer.vue (hunk 9 above)
Comment (sh16011993) [reply sources/packages/web/src/components/layouts/BodyHeaderContainer.vue]: Done

File: sources/packages/web/src/views/aest/student/Applications.vue (hunk 14 above)
Comment (sh16011993) [reply sources/packages/web/src/views/aest/student/Applications.vue]: done

File: sources/packages/web/src/components/common/OverawardDetails.vue (hunk 18 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/OverawardDetails.vue:2]: True, I forgot that the full-page-container also has the v-card optionally.

**Comment by andrepestana-aot** [comment]
> Thanks for doing the changes. The ticket is not precise about the ticket expectations but I was under the impression that we would be changing Student and Institution tabs as part of this effort, but I may be wrong. @dheepak-aot @guru-aot @andrepestana-aot @ann-aot does someone recall the final goal of the ticket?

In my understanding the main goal of the ticket is to create the component (tab-container) to make the spacing uniform among the tabs and deal with diferent layouts like overawards tab that has multiple "containers". We can keep it only for the ministry and create another ticket to apply to other cases.

**Review by dheepak-aot — COMMENTED** [review]
Thanks for doing the changes @sh16011993 . LGTM.

**Review by dheepak-aot — APPROVED** [review]

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, looks good 👍
